src/ring_theory/dedekind_domain/finite_adele_ring.lean

```diff
@@ -0,0 +1,272 @@
+field, its finite adèle ring is just defined to be the trivial ring.
+
+## References
+* [J.W.S. Cassels, A. Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+
+## Tags
+finite adèle ring, dedekind domain
+-/
+
+noncomputable theory
+open function set is_dedekind_domain is_dedekind_domain.height_one_spectrum
+
+namespace dedekind_domain
+
+variables (R K : Type*) [comm_ring R] [is_domain R] [is_dedekind_domain R] [field K]
+  [algebra R K] [is_fraction_ring R K] (v : height_one_spectrum R)
+
+/-- The product of all `adic_completion_integers`, where `v` runs over the maximal ideals of `R`. -/
+@[derive [comm_ring, topological_space, inhabited]]
+def finite_integral_adeles := Π (v : height_one_spectrum R), v.adic_completion_integers K
+
+local notation `R_hat` := finite_integral_adeles
+
+/-- The product of all `adic_completion`, where `v` runs over the maximal ideals of `R`. -/
+@[derive [non_unital_non_assoc_ring, topological_space, topological_ring, comm_ring, inhabited]]
+def prod_adic_completions := Π (v : height_one_spectrum R), v.adic_completion K
+
+local notation `K_hat` := prod_adic_completions
+
+namespace finite_integral_adeles
+
+noncomputable! instance : has_coe (R_hat R K) (K_hat R K) := { coe := λ x v, x v }
+
+lemma coe_apply (x : R_hat R K) (v : height_one_spectrum R) : (x : K_hat R K) v = ↑(x v) := rfl
+
+/-- The inclusion of `R_hat` in `K_hat` as a homomorphism of additive monoids. -/
+def coe.add_monoid_hom : add_monoid_hom (R_hat R K) (K_hat R K) :=
+{ to_fun    := coe,
+  map_zero' := rfl,
+  map_add'  := λ x y, by { ext v, simp only [coe_apply, pi.add_apply, subring.coe_add] }}
+
+lemma coe.add_monoid_hom_apply (x : R_hat R K) (v : height_one_spectrum R) :
+  (coe.add_monoid_hom R K) x v = x v := rfl
+
+/-- The inclusion of `R_hat` in `K_hat` as a ring homomorphism. -/
+def coe.ring_hom : ring_hom (R_hat R K) (K_hat R K)  :=
+{ to_fun   := coe,
+  map_one' := rfl,
+  map_mul' := λ x y, by {ext p, simp only [pi.mul_apply, subring.coe_mul], refl },
+  ..coe.add_monoid_hom R K }
+
+lemma coe.ring_hom_apply (x : R_hat R K) (v : height_one_spectrum R) :
+  (coe.ring_hom R K) x v = x v := rfl
+
+end finite_integral_adeles
+
+section algebra_instances
+
+instance : algebra K (K_hat R K) :=
+(by apply_instance : algebra K ((Π (v : height_one_spectrum R), v.adic_completion K)))
+
+instance prod_adic_completions.algebra' : algebra R (K_hat R K) :=
+(by apply_instance : algebra R (Π (v : height_one_spectrum R), v.adic_completion K))
+
+instance : is_scalar_tower R K (K_hat R K) :=
+(by apply_instance : is_scalar_tower R K ((Π (v : height_one_spectrum R), v.adic_completion K)))
+
+instance : algebra R (R_hat R K) :=
+(by apply_instance : algebra R (Π (v : height_one_spectrum R), v.adic_completion_integers K))
+
+instance prod_adic_completions.algebra_completions : algebra (R_hat R K) (K_hat R K) :=
+(finite_integral_adeles.coe.ring_hom R K).to_algebra
```

Comment:
`to_algebra` is usually a mistake as it can create diamonds. Does a smul instance for these types already exist elsewhere?

Reply:
I don't think so - `instance : has_smul (R_hat R K) (K_hat R K) := infer_instance` doesn't find anything.